leaflet-treering.js

```diff
@@ -348,19 +348,77 @@ function MeasurementData (dataObject) {
    * @function insertPoint
    */
   MeasurementData.prototype.insertPoint = function(latLng, hasLatewood) {
-    var i = 0;
-    while (this.points[i] != undefined &&
-        this.points[i].latLng.lng < latLng.lng) {
-      i++;
-    }
+    var disList = [];
+
+    /**
+    * calculate the distance between 2 points
+    * @function distanceCalc
+    * @param {first point.latLng} pointA
+    * @param {second point.latLng} pointB
+    */
+    function distanceCalc (pointA, pointB) {
+      return Math.sqrt(Math.pow((pointB.lng - pointA.lng), 2) +
+                       Math.pow((pointB.lat - pointA.lat), 2));
+    };
+
+    // finds point with smallest abs. distance
+    for (i = 0; i < this.points.length; i++) {
+      var distance = Number.MAX_SAFE_INTEGER;
+      if (this.points[i] && this.points[i].latLng) {
+         var currentPoint = this.points[i].latLng;
+         distance = distanceCalc(currentPoint, latLng);
+      disList.push(distance);
+    };
+
+    var minDistance = Math.min(...disList);
+    i = disList.indexOf(minDistance)
+
     if (this.points[i] == null) {
-      alert('New point must be within existing points.' +
-          'Use the create toolbar to add new points to the series.');
+      alert('New point must be within existing points. Use the create toolbar to add new points to the series.');
       return;
     }
 
+    // define 3 points: i, i - 1, i + 1
+    var iPoint = this.points[i].latLng;
+
+    if (this.points[i - 1].start) {
+      var iMinus = this.points[i].latLng;
+    } else {
+      var iMinus = this.points[i - 1].latLng;
+    };
+
+    if (this.points[i + 1]) { //
+      var iPlus = this.points[i + 1].latLng;
+    } else {
+      var iPlus = L.latLng(0.15, 0.6);
+    };
+
+    // distance: point[i] to point[i + 1]
+    var disIPlus = distanceCalc(iPoint, iPlus);
+    // distance: point[i} to point[i - 1]
+    var disIMinus = distanceCalc(iPoint, iMinus);
+    // distance: point[i] to inserted point
+    var disIInsert = distanceCalc(iPoint, latLng);
+    // distance: point[i + 1] to inserted point
+    var disPlusInsert = distanceCalc(iPlus, latLng);
+    // distance: point[i - 1] to inserted point
+    var disMinusInsert = distanceCalc(iMinus, latLng);
+
+    // law of cosines to find inner angle
+    var numeratorPlus = (disPlusInsert ** 2) - ((disIInsert ** 2) + (disIPlus ** 2));
+    var denominatorPlus = -2 * disIInsert * disIPlus;
+    var numeratorMinus = (disMinusInsert ** 2) - ((disIInsert ** 2) + (disIMinus ** 2));
+    var denominatorMinus = -2 * disIInsert * disIMinus;
+    var anglePlus = Math.acos(numeratorPlus/denominatorPlus);
```

Comment:
is denominatorPLus guaranteed to never be zero? If not, you should have a case for this (and the next line) to consider what will happen and avoid a divide by zero failure.

Reply:
It would only be zero if the distance between points[i] and the inserted point is 0. However upon testing that by placing a new point on top of itself several times, no errors occur.